File: Solutions/Common Problems/Design Questions/Design HashSet.py
class MyHashSet:

    def __init__(self):
        self.buckets = 1000
        self.hash = [[] for _ in range(self.buckets)]

    def findHash(self, key):
        return key % self.buckets
        
    def add(self, key: int) -> None:
        bucket = self.findHash(key)
        
        if not self.contains(key):
            self.hash[bucket].append(key)    

    def remove(self, key: int) -> None:
        bucket = self.findHash(key)
        
        if self.contains(key):
            self.hash[bucket].remove(key)

    def contains(self, key: int) -> bool:
        bucket = self.findHash(key)
        
        return key in self.hash[bucket]


# Your MyHashSet object will be instantiated and called as such:
# obj = MyHashSet()
# obj.add(key)
# obj.remove(key)
# param_3 = obj.contains(key)

File: Solutions/Common Problems/Design Questions/Design HashSet.cpp
class MyHashSet {
public:
    int buckets = 1000;  // number of buckets
    vector<list<int>> hash;  // each bucket has a linked list
    
    MyHashSet() : hash(1000) {
        
    }
    
    int findHash(int key) {
        return key % buckets;
    }
    
    void add(int key) {
        int val = findHash(key);
        
        // add key if not already present
        if (!contains(key))
            hash[val].push_back(key);
    }
    
    void remove(int key) {
        int val = findHash(key);
        
        hash[val].remove(key);
    }
    
    bool contains(int key) {
        int val = findHash(key);
        
        return find(hash[val].begin(), hash[val].end(), key) != hash[val].end();
    }
};

/*
- our hashmap consists of 1000 buckets.
- each bucket consists of a linked list of keys.
- the hash function maps each key to a bucket.
  here, we use the hash (key % num_buckets).

let k = number of buckets
time complexity:
    - add: O(n / k)
    - remove: O(n / k)
    - contains: O(n / k)

optimization: 
    - using a balanced BST (c++ set) in place of a linked 
      list will make all time complexities O(log (n / k)).
*/